src/mesh_smoothing.rs

```diff
@@ -0,0 +1,582 @@
+
+        for i in 0..test_geometry_1_i_vertices.len() {
+            assert!(
+                distance_squared(
+                    &test_geometry_1_i_vertices[i],
+                    &relaxed_geometry_1_i_vertices[i]
+                ) < TOLERANCE_SQUARED
+            );
+        }
+
+        let relaxed_geometry_3_i_vertices = relaxed_geometry_3_i.vertices();
+        let test_geometry_3_i_vertices = test_geometry_3_i.vertices();
+
+        for i in 0..test_geometry_3_i_vertices.len() {
+            assert!(
+                distance_squared(
+                    &test_geometry_3_i_vertices[i],
+                    &relaxed_geometry_3_i_vertices[i]
+                ) < TOLERANCE_SQUARED
+            );
+        }
+    }
+
+    #[test]
+    fn test_laplacian_smoothing_with_anchors() {
+        let (faces, vertices) = shape_for_smoothing_with_anchors();
+        let geometry = Geometry::from_triangle_faces_with_vertices_and_computed_normals(
+            faces.clone(),
+            vertices.clone(),
+            NormalStrategy::Sharp,
+        );
+
+        let fixed_vertex_indices: Vec<u32> = vec![0, 1, 7, 8, 6];
+
+        let (faces_correct, vertices_correct) = shape_for_smoothing_with_anchors_50_iterations();
+        let test_geometry_correct =
+            Geometry::from_triangle_faces_with_vertices_and_computed_normals(
+                faces_correct.clone(),
+                vertices_correct.clone(),
+                NormalStrategy::Sharp,
+            );
+
+        let relaxed_geometry =
+            laplacian_smoothing_with_anchors(&geometry, 50, &fixed_vertex_indices);
+
+        let relaxed_geometry_faces: Vec<TriangleFace> =
+            relaxed_geometry.triangle_faces_iter().collect();
+        let test_geometry_faces: Vec<TriangleFace> =
+            test_geometry_correct.triangle_faces_iter().collect();
+
+        assert_eq!(relaxed_geometry_faces, test_geometry_faces);
+
+        const TOLERANCE_SQUARED: f32 = 0.01 * 0.01;
+
+        let relaxed_geometry_vertices = relaxed_geometry.vertices();
+        let test_geometry_vertices = test_geometry_correct.vertices();
+
+        for i in 0..test_geometry_vertices.len() {
+            assert!(
+                distance_squared(&test_geometry_vertices[i], &relaxed_geometry_vertices[i])
+                    < TOLERANCE_SQUARED
+            );
+        }
+    }
+
+    #[test]
+    fn test_laplacian_smoothing_with_anchors_find_border_vertices() {
+        let (faces, vertices) = shape_for_smoothing_with_anchors();
+        let geometry = Geometry::from_triangle_faces_with_vertices_and_computed_normals(
+            faces.clone(),
+            vertices.clone(),
+            NormalStrategy::Sharp,
+        );
+
+        let oriented_edges: Vec<OrientedEdge> = geometry.oriented_edges_iter().collect();
+        let edge_valency_map = edge_valencies(&oriented_edges);
+        let fixed_vertex_indices: Vec<_> = border_vertex_indices(&edge_valency_map)
```

Comment:
This can be written more succinctly by using `FromItarator`.

```rust
use std::iter::FromIterator;
// ... later ...
let vector = Vec::from_iter(set.into_iter());
```